lib/block_view_validator.go

```diff
@@ -0,0 +1,2133 @@
+	var blsPublicKeyBytes []byte
+	if blsPublicKey != nil {
+		blsPublicKeyBytes = blsPublicKey.ToBytes()
+	}
+	return EncodeByteArray(blsPublicKeyBytes)
+}
+
+func DecodeBLSPublicKey(rr io.Reader) (*bls.PublicKey, error) {
+	publicKeyBytes, err := DecodeByteArray(rr)
+	if err != nil {
+		return nil, err
+	}
+	return (&bls.PublicKey{}).FromBytes(publicKeyBytes)
+}
+
+func EncodeBLSSignature(blsSignature *bls.Signature) []byte {
+	var blsSignatureBytes []byte
+	if blsSignature != nil {
+		blsSignatureBytes = blsSignature.ToBytes()
+	}
+	return EncodeByteArray(blsSignatureBytes)
+}
+
+func DecodeBLSSignature(rr io.Reader) (*bls.Signature, error) {
+	signatureBytes, err := DecodeByteArray(rr)
+	if err != nil {
+		return nil, err
+	}
+	return (&bls.Signature{}).FromBytes(signatureBytes)
+}
+
+func CreateValidatorVotingSignaturePayload(
+	transactorPublicKeyBytes []byte,
+	votingPublicKey *bls.PublicKey,
+	votingSignatureBlockHeight uint64,
+) []byte {
+	var data []byte
+	// HASH(TransactorPublicKey)
+	hashedTransactorPublicKey := sha256.Sum256(transactorPublicKeyBytes)
+	data = append(data, hashedTransactorPublicKey[:]...)
+	// HASH(VotingPublicKey)
+	hashedVotingPublicKey := sha256.Sum256(votingPublicKey.ToBytes())
+	data = append(data, hashedVotingPublicKey[:]...)
+	// VotingSignatureBlockHeight
+	data = append(data, UintToBuf(votingSignatureBlockHeight)...)
+	// HASH( HASH(TransactorPublicKey) || HASH(VotingPublicKey) || VotingSignatureBlockHeight )
+	hashedData := sha256.Sum256(data)
+	return hashedData[:]
+}
+
+//
+// CONSTANTS
+//
+
+const RuleErrorProofofStakeTxnBeforeBlockHeight RuleError = "RuleErrorProofOfStakeTxnBeforeBlockHeight"
+const RuleErrorInvalidValidatorPKID RuleError = "RuleErrorInvalidValidatorPKID"
+const RuleErrorValidatorNoDomains RuleError = "RuleErrorValidatorNoDomains"
+const RuleErrorValidatorTooManyDomains RuleError = "RuleErrorValidatorTooManyDomains"
+const RuleErrorValidatorInvalidDomain RuleError = "RuleErrorValidatorInvalidDomain"
+const RuleErrorValidatorDuplicateDomains RuleError = "RuleErrorValidatorDuplicateDomains"
+const RuleErrorValidatorNotFound RuleError = "RuleErrorValidatorNotFound"
+const RuleErrorValidatorMissingVotingPublicKey RuleError = "RuleErrorValidatorMissingVotingPublicKey"
+const RuleErrorValidatorMissingVotingPublicKeySignature RuleError = "RuleErrorValidatorMissingVotingPublicKeySignature"
+const RuleErrorValidatorInvalidVotingPublicKeySignature RuleError = "RuleErrorValidatorInvalidVotingPublicKeySignature"
+const RuleErrorValidatorInvalidVotingSignatureBlockHeight RuleError = "RuleErrorValidatorInvalidVotingSignatureBlockHeight"
+const RuleErrorValidatorDisablingExistingDelegatedStakers RuleError = "RuleErrorValidatorDisablingExistingDelegatedStakers"
+const RuleErrorUnjailingNonjailedValidator RuleError = "RuleErrorUnjailingNonjailedValidator"
+const RuleErrorUnjailingValidatorTooEarly RuleError = "RuleErrorUnjailingValidatorTooEarly"
+
+const MaxValidatorNumDomains int = 12
```

Comment:
Hmm... I'm not sure we need a protocol-level limit on number of domains. Is there a downside to just making thi sunlimited?

Reply:
I ended up bumping it to 100 as a compromise. Can revisit later.